Optimize Integer formatting/parsing

arthurprs opened the pull request:
I wrote a reasonably optimized version for both functions. Further optimizations are possible but I tried to keep the code size small (which I think is important), it's a road of diminished gains.

The repository used for testing/benchmarks is https://github.com/arthurprs/rust-optimized-inttostr

Benchmarks are ran for 3 different distributions, bellow are string length histograms for the u32 type

* _h (big numbers skew)
[0, 0, 5, 29, 103, 212, 551, 1138, 1887, 3196, 2879]
* _m (slight small number skew):
[0, 2807, 1334, 1057, 905, 821, 772, 707, 627, 605, 365]
* _l (small numbers skew):
[0, 8004, 567, 351, 248, 212, 170, 126, 136, 112, 74]

Tested processors are
* x64 laptop (i7-2670QM)
* x32 server (Digital Ocean E5-2630L-v2)

### Display

It uses a small look up table (200 bytes) and decode up to 4 characters at a time. I also took special precautions to reduce 64bit arithmetic on 32bit architectures and the gains are huge in these cases.

Overall, on modern 64bit CPUs it's pretty much the same speed as the stdlib implementation for very small numbers (0..99), but pulls ahead as the length of the decimal increases. On slight older CPUs (w/ worse ALUs) or 32bit architectures it's pretty much always faster.

x64 benchmarks
```
test bench::display_h_new_u08     ... bench:      71,041 ns/iter (+/- 2,894)
test bench::display_h_new_u16     ... bench:     378,255 ns/iter (+/- 36,547)
test bench::display_h_new_u32     ... bench:   4,232,483 ns/iter (+/- 509,661)
test bench::display_h_new_u64     ... bench:   5,166,740 ns/iter (+/- 421,124)
test bench::display_h_stdlib_u08  ... bench:      73,536 ns/iter (+/- 5,287)
test bench::display_h_stdlib_u16  ... bench:     451,443 ns/iter (+/- 16,879)
test bench::display_h_stdlib_u32  ... bench:   5,551,070 ns/iter (+/- 518,151)
test bench::display_h_stdlib_u64  ... bench:   8,624,374 ns/iter (+/- 643,701)
test bench::display_l_new_u08     ... bench:      71,547 ns/iter (+/- 504)
test bench::display_l_new_u16     ... bench:     399,727 ns/iter (+/- 28,030)
test bench::display_l_new_u32     ... bench:   4,365,303 ns/iter (+/- 414,414)
test bench::display_l_new_u64     ... bench:   5,302,382 ns/iter (+/- 292,324)
test bench::display_l_stdlib_u08  ... bench:      75,445 ns/iter (+/- 2,487)
test bench::display_l_stdlib_u16  ... bench:     444,313 ns/iter (+/- 16,203)
test bench::display_l_stdlib_u32  ... bench:   5,761,801 ns/iter (+/- 387,186)
test bench::display_l_stdlib_u64  ... bench:   8,790,365 ns/iter (+/- 614,846)
test bench::display_m_new_u08     ... bench:      71,820 ns/iter (+/- 2,956)
test bench::display_m_new_u16     ... bench:     399,649 ns/iter (+/- 20,643)
test bench::display_m_new_u32     ... bench:   4,355,561 ns/iter (+/- 179,189)
test bench::display_m_new_u64     ... bench:   5,070,594 ns/iter (+/- 341,950)
test bench::display_m_stdlib_u08  ... bench:      74,900 ns/iter (+/- 1,909)
test bench::display_m_stdlib_u16  ... bench:     448,788 ns/iter (+/- 20,791)
test bench::display_m_stdlib_u32  ... bench:   5,717,939 ns/iter (+/- 316,824)
test bench::display_m_stdlib_u64  ... bench:   8,787,160 ns/iter (+/- 482,864)
```

x86 benchmarks
```
test bench::display_h_new_u08     ... bench:      94,246 ns/iter (+/- 34,872)
test bench::display_h_new_u16     ... bench:     533,805 ns/iter (+/- 22,499)
test bench::display_h_new_u32     ... bench:   6,127,747 ns/iter (+/- 2,192,789)
test bench::display_h_new_u64     ... bench:  14,994,203 ns/iter (+/- 1,609,345)
test bench::display_h_stdlib_u08  ... bench:     107,233 ns/iter (+/- 8,571)
test bench::display_h_stdlib_u16  ... bench:     631,186 ns/iter (+/- 11,332)
test bench::display_h_stdlib_u32  ... bench:   7,696,344 ns/iter (+/- 957,917)
test bench::display_h_stdlib_u64  ... bench:  45,677,401 ns/iter (+/- 4,991,344)
test bench::display_l_new_u08     ... bench:      95,855 ns/iter (+/- 27,735)
test bench::display_l_new_u16     ... bench:     532,084 ns/iter (+/- 40,479)
test bench::display_l_new_u32     ... bench:   5,973,953 ns/iter (+/- 211,676)
test bench::display_l_new_u64     ... bench:  14,773,064 ns/iter (+/- 1,276,579)
test bench::display_l_stdlib_u08  ... bench:     106,350 ns/iter (+/- 63,963)
test bench::display_l_stdlib_u16  ... bench:     637,746 ns/iter (+/- 101,005)
test bench::display_l_stdlib_u32  ... bench:   7,740,640 ns/iter (+/- 848,478)
test bench::display_l_stdlib_u64  ... bench:  44,846,932 ns/iter (+/- 4,514,694)
test bench::display_m_new_u08     ... bench:      94,549 ns/iter (+/- 13,029)
test bench::display_m_new_u16     ... bench:     546,030 ns/iter (+/- 35,804)
test bench::display_m_new_u32     ... bench:   5,983,924 ns/iter (+/- 1,180,559)
test bench::display_m_new_u64     ... bench:  14,817,873 ns/iter (+/- 2,271,464)
test bench::display_m_stdlib_u08  ... bench:     107,806 ns/iter (+/- 8,805)
test bench::display_m_stdlib_u16  ... bench:     630,714 ns/iter (+/- 6,586)
test bench::display_m_stdlib_u32  ... bench:   7,784,210 ns/iter (+/- 358,601)
test bench::display_m_stdlib_u64  ... bench:  46,223,927 ns/iter (+/- 6,553,176)
```

### from_str_radix (FromStr)

All valid digits are ascii so I modified the function to use the underlining bytes instead and simplified the match to avoid wasting cycles.

x64 benchmarks
```
test bench::from_str_h_new_u08    ... bench:      28,153 ns/iter (+/- 624)
test bench::from_str_h_new_u16    ... bench:     223,513 ns/iter (+/- 11,554)
test bench::from_str_h_new_u32    ... bench:   3,098,935 ns/iter (+/- 231,022)
test bench::from_str_h_new_u64    ... bench:   5,009,900 ns/iter (+/- 341,961)
test bench::from_str_h_stdlib_u08 ... bench:      34,033 ns/iter (+/- 2,068)
test bench::from_str_h_stdlib_u16 ... bench:     248,785 ns/iter (+/- 14,208)
test bench::from_str_h_stdlib_u32 ... bench:   4,150,536 ns/iter (+/- 266,070)
test bench::from_str_h_stdlib_u64 ... bench:   6,817,997 ns/iter (+/- 449,838)
test bench::from_str_l_new_u08    ... bench:      27,552 ns/iter (+/- 1,500)
test bench::from_str_l_new_u16    ... bench:     234,360 ns/iter (+/- 13,144)
test bench::from_str_l_new_u32    ... bench:   3,140,261 ns/iter (+/- 248,175)
test bench::from_str_l_new_u64    ... bench:   5,176,583 ns/iter (+/- 350,416)
test bench::from_str_l_stdlib_u08 ... bench:      35,060 ns/iter (+/- 2,154)
test bench::from_str_l_stdlib_u16 ... bench:     252,135 ns/iter (+/- 23,461)
test bench::from_str_l_stdlib_u32 ... bench:   4,154,599 ns/iter (+/- 369,606)
test bench::from_str_l_stdlib_u64 ... bench:   6,892,767 ns/iter (+/- 213,030)
test bench::from_str_m_new_u08    ... bench:      28,252 ns/iter (+/- 1,384)
test bench::from_str_m_new_u16    ... bench:     231,051 ns/iter (+/- 16,540)
test bench::from_str_m_new_u32    ... bench:   3,166,504 ns/iter (+/- 134,418)
test bench::from_str_m_new_u64    ... bench:   5,103,195 ns/iter (+/- 218,912)
test bench::from_str_m_stdlib_u08 ... bench:      35,012 ns/iter (+/- 2,735)
test bench::from_str_m_stdlib_u16 ... bench:     250,967 ns/iter (+/- 14,708)
test bench::from_str_m_stdlib_u32 ... bench:   4,101,845 ns/iter (+/- 205,802)
test bench::from_str_m_stdlib_u64 ... bench:   6,823,001 ns/iter (+/- 267,215)
```

x86 benchmarks
```
test bench::from_str_h_new_u08    ... bench:      23,682 ns/iter (+/- 3,590)
test bench::from_str_h_new_u16    ... bench:     190,916 ns/iter (+/- 29,688)
test bench::from_str_h_new_u32    ... bench:   2,649,952 ns/iter (+/- 308,576)
test bench::from_str_h_new_u64    ... bench:  23,458,434 ns/iter (+/- 2,327,427)
test bench::from_str_h_stdlib_u08 ... bench:      45,551 ns/iter (+/- 6,968)
test bench::from_str_h_stdlib_u16 ... bench:     313,739 ns/iter (+/- 17,175)
test bench::from_str_h_stdlib_u32 ... bench:   4,615,669 ns/iter (+/- 470,766)
test bench::from_str_h_stdlib_u64 ... bench:  30,589,482 ns/iter (+/- 2,278,996)
test bench::from_str_l_new_u08    ... bench:      23,763 ns/iter (+/- 5,545)
test bench::from_str_l_new_u16    ... bench:     185,472 ns/iter (+/- 33,097)
test bench::from_str_l_new_u32    ... bench:   2,691,307 ns/iter (+/- 473,886)
test bench::from_str_l_new_u64    ... bench:  22,952,593 ns/iter (+/- 1,963,742)
test bench::from_str_l_stdlib_u08 ... bench:      45,285 ns/iter (+/- 16,337)
test bench::from_str_l_stdlib_u16 ... bench:     313,624 ns/iter (+/- 6,643)
test bench::from_str_l_stdlib_u32 ... bench:   4,595,679 ns/iter (+/- 1,876,361)
test bench::from_str_l_stdlib_u64 ... bench:  30,434,683 ns/iter (+/- 1,901,996)
test bench::from_str_m_new_u08    ... bench:      23,812 ns/iter (+/- 1,505)
test bench::from_str_m_new_u16    ... bench:     185,553 ns/iter (+/- 19,788)
test bench::from_str_m_new_u32    ... bench:   2,614,920 ns/iter (+/- 66,230)
test bench::from_str_m_new_u64    ... bench:  23,241,778 ns/iter (+/- 3,474,077)
test bench::from_str_m_stdlib_u08 ... bench:      45,634 ns/iter (+/- 1,436)
test bench::from_str_m_stdlib_u16 ... bench:     316,479 ns/iter (+/- 21,212)
test bench::from_str_m_stdlib_u32 ... bench:   4,609,147 ns/iter (+/- 487,068)
test bench::from_str_m_stdlib_u64 ... bench:  30,165,173 ns/iter (+/- 1,601,830)
```

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @huonw (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

Gankro commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +238,82 @@ integer! { i8, u8 }
 integer! { i64, u64 }
+
+const DEC_DIGITS_LUT: [u8; 200] = [
+    '0' as u8,'0' as u8,'0' as u8,'1' as u8,'0' as u8,'2' as u8,'0' as u8,'3' as u8,'0' as u8,'4' as u8,'0' as u8,'5' as u8,'0' as u8,'6' as u8,'0' as u8,'7' as u8,'0' as u8,'8' as u8,'0' as u8,'9' as u8,
```
Is this something a macro could make...?

Gankro replied:
At very least, I feel like you probably want:

```
static DEC_DIGIT_LUT: &'static [u8; 200] = b"00010203...";
``

Gankro replied:
regardless all of these lines fail the style lint (> 100 lines)

Gankro commented:
So I know we went through a lot of effort to basically get top-of-the-line industrial float formatting. I imagine int formatting is a *bit* easier, but I'd be remiss to not ask: are there standard "optimal" algorithms that we should be investing in for this?

arthurprs commented:
@Gankro what triggered me to do this work was this repository https://github.com/miloyip/itoa-benchmark which is a good summary of what's out there.

What I have in this PR can be seen as a hybrid of the lut and branchlut versions. My version tries to keep code size small and avoid regressions for small numbers. In addition to that it also reduces 64 arithmetic.

Also note that the biggest increase in speed from lut to branchlut/countlut in that repo are due to avoiding the double buffering, which we can't really avoid in rust due to the fmt machinery.

Gankro commented:
CC @lifthrasiir

Gankro commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +238,82 @@ integer! { i8, u8 }
+        #[allow(unused_comparisons)]
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            let is_positive = *self >= 0;
+            let mut n = if ! is_positive {
```
style nit, we put the ! right up in the condition's grill

Gankro replied:
Also presumably you could kill the negation and just reorder the branches.

arthurprs replied:
Will do

Gankro commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +238,82 @@ integer! { i8, u8 }
+macro_rules! impl_Display {
+    ($($t:ident),*: $conv_fn:ident) => ($(
+    impl fmt::Display for $t {
+        #[allow(unused_comparisons)]
```
? Is this because you put is_positive in a variable?

arthurprs replied:
Actually it's not needed anymore, I will remove it.

tamird replied:
should this have been removed?

Gankro replied:
Ah, good catch!

arthurprs replied:
Oh I forgot to leave a note about that line.

It's required for unsigned expansions, because is_positive always evaluate to true. Otherwise the compiler will complain.

Gankro commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +238,82 @@ integer! { i8, u8 }
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            let is_positive = *self >= 0;
+            let mut n = if ! is_positive {
+                let Wrapping(n) = Wrapping(!self.$conv_fn()) + Wrapping(1);
```
(!self.$conv_fn()).wrapping_add(1) also works

Gankro replied:
Also a comment as to why you do this would be great.

arthurprs replied:
Coll, will integrate that.

Gankro commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +238,82 @@ integer! { i8, u8 }
+                        let d2 = (rem % 100) << 1;
+                        curr -= 4;
+                        ptr::copy_nonoverlapping(lut_ptr.offset(d1), buf_ptr.offset(curr), 2);
+                        ptr::copy_nonoverlapping(lut_ptr.offset(d2), buf_ptr.offset(curr + 2), 2);
```
Are copy_nonoverlapping's necessary here? Surely

```
buf[curr] = DEC_DIGITS_LUT[d1]; 
buf[curr + 1] = DEC_DIGITS_LUT[d1 + 1]; 
```

Is sufficient. Was there a profile'd difference? (not that you don't need to worry about writing uninit memory since u8 isn't Drop.

arthurprs replied:
copy_* optimizes to a "mov word" and makes sure compiler avoids a bounds check.

Gankro replied:
I know it will optimize great, I'm just wondering if we need to use unsafe code to guarantee that. As in, does it *not* optimize well without this?

arthurprs replied:
Using 4 array assignments I get

```
	mov	al, byte ptr [r14 + 2*rdx]
	mov	byte ptr [rbp + r15 - 52], al
	mov	al, byte ptr [r14 + 2*rdx + 1]
	mov	byte ptr [rbp + r15 - 51], al
	mov	al, byte ptr [r14 + 2*rbx]
	mov	byte ptr [rbp + r15 - 50], al
	mov	al, byte ptr [r14 + 2*rbx + 1]
	mov	byte ptr [rbp + r15 - 49], al
```

Using the 2 copy*

```
	mov	ax, word ptr [r14 + 2*rdx]
	mov	word ptr [rbp + r15 - 52], ax
	mov	ax, word ptr [r14 + 2*rbx]
	mov	word ptr [rbp + r15 - 50], ax
```

arthurprs replied:
Using 
```
buf[(curr + 0) as usize] = DEC_DIGITS_LUT[d1 as usize]; 
buf[(curr + 1) as usize] = DEC_DIGITS_LUT[(d1 + 1) as usize]; 
buf[(curr + 2) as usize] = DEC_DIGITS_LUT[d2 as usize]; 
buf[(curr + 3) as usize] = DEC_DIGITS_LUT[(d2 + 1) as usize]; 
```

```
	cmp	rax, 199
	ja	.LBB21_16
	cmp	r8, 19
	ja	.LBB21_18
	mov	bl, byte ptr [rax + r13]
	mov	byte ptr [rbp + r8 - 64], bl
	or	rax, 1
	cmp	rax, 199
	ja	.LBB21_19
	lea	r10, [r8 + 1]
	cmp	r10, 19
	ja	.LBB21_20
	shr	rdx, 2
	imul	rdx, rdx, 100
	sub	rcx, rdx
	add	rcx, rcx
	mov	al, byte ptr [rax + r13]
	mov	byte ptr [rbp + r8 - 63], al
	cmp	rcx, 199
	ja	.LBB21_21
	lea	rax, [r8 + 2]
	cmp	rax, 19
	ja	.LBB21_23
	mov	al, byte ptr [rcx + r13]
	mov	byte ptr [rbp + r8 - 62], al
	or	rcx, 1
	cmp	rcx, 199
	ja	.LBB21_25
	lea	rax, [r8 + 3]
	cmp	rax, 19
	ja	.LBB21_26
	mov	al, byte ptr [rcx + r13]
	mov	byte ptr [rbp + r8 - 61], al
```

Gankro replied:
Woof. Alright, I approve this use of unsafe.
:100:

Gankro commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +238,82 @@ integer! { i8, u8 }
+            } else {
+                self.$conv_fn()
+            };
+            let mut buf: [u8; 20] = unsafe { mem::uninitialized() };
```
I would also be interested to know if this makes a difference over [0; 20].

arthurprs replied:
It saves the following

```
# zero sse register
	xorps	xmm0, xmm0
# zero 16 bytes using sse register
	movaps	xmmword ptr [rbp - 64], xmm0
# zero the last 4 bytes
	mov	dword ptr [rbp - 48], 0
```

Gankro replied:
:+1:

Gankro commented on src/libcore/num/mod.rs:
```diff
@@ -1446,21 +1447,29 @@ doit! { i8 i16 i32 i64 isize u8 u16 u32 u64 usize }
 
 fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32)
                                          -> Result<T, ParseIntError> {
+    const MINUS: u8 = '-' as u8;
```
`b'-'`

Gankro commented on src/libcore/num/mod.rs:
```diff
@@ -1446,21 +1447,29 @@ doit! { i8 i16 i32 i64 isize u8 u16 u32 u64 usize }
-            for c in src.chars() {
-                let x = match c.to_digit(radix) {
+            for &c in src {
+                let x = match (c as char).to_digit(radix) {
```
cheaper, I guess?

arthurprs replied:
Yes.

eefriedman replied:
Please write a wrapper function with a comment to perform the operation `(c as char).to_digit(radix)`.  The fact that `char::to_digit` returns None for Unicode scalar values between 128 and 255 isn't immediately obvious.

Gankro commented on src/libcore/num/mod.rs:
```diff
@@ -1446,21 +1447,29 @@ doit! { i8 i16 i32 i64 isize u8 u16 u32 u64 usize }
+    // all valid digits are ascii
+    let src = src.as_bytes();
+    match (src[0], &src[1..])  {
+        (MINUS, src) if src.is_empty() => Err(PIE { kind: Empty }),
```
Surely the program will panic due to indexing out of bounds before ever taking this branch?

Gankro replied:
In which case this is probably better structured as something other than a match.

arthurprs replied:
There's an is_empty() above to avoid OOB.

Gankro replied:
Ah yes; why are you checking again?

Gankro replied:
No wait I'm stupid. Too many variables called `src`.

Gankro replied:
Might be worth using a different name (e.g. digits).

Gankro commented:
I'm pretty sure this merits some new tests for interesting corner-cases.

arthurprs commented:
@Gankro thank you very much for your review. I cleaned up the code to account for your suggestions and removed the [last bounds checking](https://github.com/arthurprs/rust-optimized-inttostr/commit/3c4c7e072249f1781fd7a445b9e28c6319ccc768).

As for new tests I'm not aware of any border cases other than non-digits characters (added) and overflows (already present).

Gankro commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +240,78 @@ integer! { i8, u8 }
+            }
+
+            let buf_slice = unsafe {
+                str::from_utf8_unchecked(slice::from_raw_parts(buf_ptr.offset(curr)), buf.len() - curr as usize)
```
Looks like you got your parens wrong here

Gankro replied:
Also this line is too long (see travis)

arthurprs commented:
adjusted the nits and parens, travis is green.

Gankro commented:
This is looking great! The only outstanding issue is @eefriedman's concern about the char conversion, which I don't have strong feelings about.

arthurprs commented:
I can't find @eefriedman comment here, just in my inbox.

to_digit returns None if the char can't be converted to a number of the specified radix, it sounds simple enough to me. If you guys want I can add the wrapper function with some comments though.

eefriedman commented:
GitHub collapses comments against old changes; that's probably why you're not finding it.

The problem isn't that to_digit is weird, it's that converting a UTF-8 byte to a Rust `char` is weird.  Might be enough to just expand the "all valid digits are ASCII" comment a bit.

arthurprs commented:
Updated comment

```
    // all valid digits are ascii, so we will just iterate over the utf8 bytes
    // and cast them to chars. .to_digit() will safely return None for anything
    // other than a valid ascii digit a the given radix, including the first-byte
    // of multi-byte sequences.
```

Gankro commented:
@bors r+

Awesome stuff!

bors commented:
:pushpin: Commit 647f87f has been approved by `Gankro`

<!-- @bors r=Gankro 647f87fa8f1cc2e31d07edc27b64dbcc1b1519cb -->

bors commented:
:hourglass: Testing commit 647f87f with merge b806a4c...

tamird commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +240,80 @@ integer! { i8, u8 }
+                }
+
+                // if we reach here numbers are <= 9999, so at most 4 chars long
+                let mut n = n as isize; // possibly reduce 64bit math
```
should this be i64 rather than isize?

tamird replied:
nm, i think this is ok

Gankro replied:
Yeah it's fine specifically because we know n <= 9999

tamird commented on src/libcore/fmt/num.rs:
```diff
@@ -233,3 +240,80 @@ integer! { i8, u8 }
+                (!self.$conv_fn()).wrapping_add(1)
+            };
+            let mut buf: [u8; 20] = unsafe { mem::uninitialized() };
+            let mut curr = buf.len() as isize;
```
hm, should these be unsized?

Gankro replied:
What do you mean?

tamird replied:
ah, i thought this was a position in the buffer but it looks like it's an offset? probably fine

Gankro replied:
Yeah ptr::offset wants signed. No concern of overflow or anything.

Gankro commented:
Ehh the build already started. Unless it bounces spuriously (which is not unheard of), just submit a followup pr to remove the annotation.

Such is the danger of weekend PRs... no queue!

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5737)

Gankro commented:
Non spurious failures even:

```

rustc: x86_64-apple-darwin/stage2/test/coretesttest-x86_64-apple-darwin
../src/libcoretest/num/mod.rs:121:47: 121:60 error: `ParseIntError` does not name a structure [E0422]
../src/libcoretest/num/mod.rs:121         assert_eq!("--129".parse::<i8>(), Err(ParseIntError{ kind: IntErrorKind::InvalidDigit }));
                                                                                ^~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:121:9: 121:99 note: expansion site
../src/libcoretest/num/mod.rs:121:68: 121:94 error: failed to resolve. Use of undeclared type or module `IntErrorKind` [E0433]
../src/libcoretest/num/mod.rs:121         assert_eq!("--129".parse::<i8>(), Err(ParseIntError{ kind: IntErrorKind::InvalidDigit }));
                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:121:9: 121:99 note: expansion site
../src/libcoretest/num/mod.rs:121:68: 121:94 error: unresolved name `IntErrorKind::InvalidDigit` [E0425]
../src/libcoretest/num/mod.rs:121         assert_eq!("--129".parse::<i8>(), Err(ParseIntError{ kind: IntErrorKind::InvalidDigit }));
                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:121:9: 121:99 note: expansion site
../src/libcoretest/num/mod.rs:122:47: 122:60 error: `ParseIntError` does not name a structure [E0422]
../src/libcoretest/num/mod.rs:122         assert_eq!("Съешь".parse::<u8>(), Err(ParseIntError{ kind: IntErrorKind::InvalidDigit }));
                                                                                ^~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:122:9: 122:99 note: expansion site
../src/libcoretest/num/mod.rs:122:68: 122:94 error: failed to resolve. Use of undeclared type or module `IntErrorKind` [E0433]
../src/libcoretest/num/mod.rs:122         assert_eq!("Съешь".parse::<u8>(), Err(ParseIntError{ kind: IntErrorKind::InvalidDigit }));
                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:122:9: 122:99 note: expansion site
../src/libcoretest/num/mod.rs:122:68: 122:94 error: unresolved name `IntErrorKind::InvalidDigit` [E0425]
../src/libcoretest/num/mod.rs:122         assert_eq!("Съешь".parse::<u8>(), Err(ParseIntError{ kind: IntErrorKind::InvalidDigit }));
                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:122:9: 122:99 note: expansion site
../src/libcoretest/num/mod.rs:127:43: 127:56 error: `ParseIntError` does not name a structure [E0422]
../src/libcoretest/num/mod.rs:127         assert_eq!("-".parse::<i8>(), Err(ParseIntError{ kind: IntErrorKind::Empty }));
                                                                            ^~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:127:9: 127:88 note: expansion site
../src/libcoretest/num/mod.rs:127:64: 127:83 error: failed to resolve. Use of undeclared type or module `IntErrorKind` [E0433]
../src/libcoretest/num/mod.rs:127         assert_eq!("-".parse::<i8>(), Err(ParseIntError{ kind: IntErrorKind::Empty }));
                                                                                                 ^~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:127:9: 127:88 note: expansion site
../src/libcoretest/num/mod.rs:127:64: 127:83 error: unresolved name `IntErrorKind::Empty` [E0425]
../src/libcoretest/num/mod.rs:127         assert_eq!("-".parse::<i8>(), Err(ParseIntError{ kind: IntErrorKind::Empty }));
                                                                                                 ^~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:127:9: 127:88 note: expansion site
../src/libcoretest/num/mod.rs:128:42: 128:55 error: `ParseIntError` does not name a structure [E0422]
../src/libcoretest/num/mod.rs:128         assert_eq!("".parse::<u8>(), Err(ParseIntError{ kind: IntErrorKind::Empty }));
                                                                           ^~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:128:9: 128:87 note: expansion site
../src/libcoretest/num/mod.rs:128:63: 128:82 error: failed to resolve. Use of undeclared type or module `IntErrorKind` [E0433]
../src/libcoretest/num/mod.rs:128         assert_eq!("".parse::<u8>(), Err(ParseIntError{ kind: IntErrorKind::Empty }));
                                                                                                ^~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:128:9: 128:87 note: expansion site
../src/libcoretest/num/mod.rs:128:63: 128:82 error: unresolved name `IntErrorKind::Empty` [E0425]
../src/libcoretest/num/mod.rs:128         assert_eq!("".parse::<u8>(), Err(ParseIntError{ kind: IntErrorKind::Empty }));
                                                                                                ^~~~~~~~~~~~~~~~~~~
<std macros>:1:1: 9:39 note: in expansion of assert_eq!
../src/libcoretest/num/mod.rs:128:9: 128:87 note: expansion site
error: aborting due to 12 previous errors
make: *** [x86_64-apple-darwin/stage2/test/coretesttest-x86_64-apple-darwin] Error 101
```

arthurprs commented:
Fixed. Sorry, I screwed up the tests, I have to get a better machine so I can compile/test rustc in doable time.

Gankro commented:
No worries!

We all but corners testing PRs. :D

@bors r+

bors commented:
:pushpin: Commit c073f81 has been approved by `Gankro`

<!-- @bors r=Gankro c073f81920bc2324b026cf9610a096df0d97c6a5 -->

bors commented:
:hourglass: Testing commit c073f81 with merge a41178e...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/5753), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/5768), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/5710), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/5739), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5715), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/5790), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/5781), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5746), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/726), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/729), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/724), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/728)
